### ISSUE TITLE
[Bug] xgrammar structured output: array<string> without maxItems expands indefinitely (repetition until max_tokens) on Qwen3-235B

### ISSUE THREAD

**Issue by linnea-lin-00638949** [vllm-project/vllm-ascend#13646]
### Summary
When using vLLM structured output (`response_format: json_schema`) with a schema containing an `array<string>` field **without `maxItems`**, the model expands the array indefinitely — generating hundreds-to-thousands of items until `max_tokens` is hit (`finish_reason: length`). After the unique-item vocabulary is exhausted, output degrades into two-item alternating repetition. This reproduces on vLLM-Ascend 0.24.0 (nightly-main) with Qwen3-235B-A22B-Instruct (w8a8), 8x Ascend 910A3, TP=8, xgrammar backend.

### Reproduction

**Request** (loose schema, `required` only `name`/`age`, no `strict`, no `maxItems`):
```json
{"model":"auto","messages":[{"role":"user","content":"请生成一个包含多种类型字段的数据。"}],
 "response_format":{"type":"json_schema","json_schema":{"name":"multi_type_data",
   "schema":{"type":"object","properties":{"name":{"type":"string"},"age":{"type":"integer"},
     "score":{"type":"number"},"is_active":{"type":"boolean"},
     "tags":{"type":"array","items":{"type":"string"}}},"required":["name","age"]}}},
 "stream":false,"max_tokens":4096}
```

**Environment**
- Image: vllm-ascend nightly-main (vllm 0.24.0, vllm-ascend commit e1bd15041)
- Model: Qwen3-235B-A22B-Instruct-2507, w8a8 (ascend quantization)
- Hardware: 8x Ascend 910A3, TP=8
- `--structured-outputs-config '{"backend":"xgrammar","disable_any_whitespace":true}'`
- `--quantization ascend`, `--max-num-seqs 384`, `--gpu-memory-utilization 0.9`

**Results** (max_tokens=300, 5 shots, loose schema)
| shot | finish | completion_tokens | tag_items | expand |
|---|---|---|---|---|
| 1 | length | 300 | 64 | ✅ |
| 2 | length | 300 | 33 | ✅ |
| 3 | length | 300 | 59 | ✅ |
| 4 | length | 300 | 71 | ✅ |
| 5 | length | 300 | 63 | ✅ |

5/5 = 100% expansion. Strict schema (`strict:true` + all required + `additionalProperties:false`): 2/3 = 66% expansion.

**Long run** (max_tokens=4096, 1 shot): `finish=length`, 4096 tokens, **1118 tag items, 135 unique, 983 duplicates (88% dup)**. Head expands distinct labels (`"用户","VIP","北京","技术爱好者",...`); tail still expanding (`"专家","学者","研究员",...`).

Online production (max_tokens=51200): 51200 tokens, **10271 items, 204 unique, 98% dup** — after 204 distinct items, degrades into two-item alternating repetition (`"支持与陪伴","陪伴与支持","支持与陪伴",...`).

### Root cause (analysis)
- xgrammar compiles `array<string>` (no `maxItems`) to BNF `[item (, item)*]` — array length is unbounded.
- At each item boundary, the grammar mask allows **both** `,` (continue) and `]`/`]}` (close). The close path is reachable.
- However, model logits strongly prefer `,` (continue, logprob ≈ 0) over the close token (logprob -2 ~ -15). The model keeps selecting `,` → infinite expansion.
- plain mode (no `response_format`) stops at ~3 items (close-token logprob rises to 0). Structured mode changes the close-point logits via grammar-mask cascade (different token choices → different KV cache → close logprob rises too slowly / non-monotonically, never exceeds `,`).
- This is **not** a mask-precision issue, **not** a forbidden-close-token issue (close tokens are in the allowed set), and **not** fixed by `disable_any_whitespace` (compact mode only affects whitespace, not array cardinality).

### Workarounds
- **`maxItems` in schema** (e.g. `tags: {"type":"array","maxItems":5}`): 0% expansion — xgrammar enforces array length at the BNF level. Confirmed working.
- `disable_any_whitespace=true` does **not** fix array expansion (it does fix the related `integer` age-repetition issue).

### Related
- vllm-project/vllm#40099 / #40097 — auto-enable repetition detection for grammar-constrained structured output (open)
- vllm-project/vllm#44619 — limit whitespace in JSON schema FSM (open, not merged)
- vllm-project/vllm#17865 — xgrammar minItems/maxItems support (only enforced when schema declares maxItems)
- mlc-ai/xgrammar#160 — array minItems/maxItems support
- mlc-ai/xgrammar#802 — grammar runaway loop (same mechanism)
- vllm-project/vllm-ascend#5835 — Qwen3-235B + json_schema output issue (different symptom: garbled chars)
- vllm-project/vllm-ascend#8328 — structured output hang under certain Ascend config

### Suggested fix direction
1. Auto-enable repetition/loop detection for grammar-constrained requests (vllm#40099 direction).
2. xgrammar: default upper bound for unbounded array/string repetition (per-state token budget, xgrammar#688 direction).
3. At minimum: document `array<string>` + no `maxItems` + large `max_tokens` as a dangerous combination.

**Comment by linnea-lin-00638949** [comment]
## 复现细节补充（启动脚本 / curl / 权重）

### 启动脚本
```bash
export VLLM_VERSION=0.24.0          # 必须与镜像 vllm 实际版本一致
export MODEL_PATH=/home/admin/model/
vllm serve ${MODEL_PATH} \
  --host 0.0.0.0 --port 8000 --trust-remote-code --served-model-name auto \
  --distributed-executor-backend mp -tp=8 \
  --model-loader-extra-config '{"enable_multithread_load":true,"num_threads":8}' \
  --max-num-seqs 384 --max-num-batched-tokens 16384 --block-size 128 \
  --no-async-scheduling --gpu-memory-utilization 0.9 --quantization ascend \
  --compilation-config '{"cudagraph_capture_sizes":[1,2,4,6,8,12,16,24,32,36,40,44,48,56,60,64,72,90,96,108,120,132,144],"cudagraph_mode":"FULL_DECODE_ONLY"}' \
  --structured-outputs-config '{"backend":"xgrammar","disable_any_whitespace":true}'
```

### 复现 curl（宽松 schema，无 strict / 无 maxItems）
```bash
curl -s -X POST http://<pod-ip>:8000/v1/chat/completions -H 'Content-Type: application/json' -d '{
  "model":"auto",
  "messages":[{"role":"user","content":"请生成一个包含多种类型字段的数据。"}],
  "response_format":{"type":"json_schema","json_schema":{"name":"multi_type_data","schema":{
    "type":"object","properties":{"name":{"type":"string"},"age":{"type":"integer"},
      "score":{"type":"number"},"is_active":{"type":"boolean"},
      "tags":{"type":"array","items":{"type":"string"}}},"required":["name","age"]}}},
  "stream":false,"max_tokens":4096}'
```

### 权重 / 镜像
- 模型: `qwen3-235b-a22b-instruct-2507`, w8a8 (Ascend modelslim 量化)
- modelhub: `modelhub_16100032_w8a8-qwen3-235b-a22b-instruct:86300084_20250926152833` (221G)
- 镜像: `hcr.meta-wulan01.hw-wulan.local/antsys/vllm:nightly-main-a3-openeuler-20260719151717_aarch64` (vllm 0.24.0 / vllm-ascend e1bd15041)
- 硬件: 8x Ascend 910A3, TP=8

> 注: 此镜像 vllm/vllm-ascend 提交配对不一致, 跑通需 2 处源码 patch (transformers 5.13 register API + spec_decode 函数改名) + VLLM_VERSION=0.24.0。正在用最新 `nightly-main-a3-openeuler` tag + BF16 权重重新复现, 结果后续更新。

**Comment by linnea-lin-00638949** [comment]
## 更新:BF16 + 最新 nightly-main 镜像复现（canonical，无需 patch）

改用 BF16 权重 + 最新 `nightly-main-a3-openeuler` 镜像(vllm **0.26.0**, vllm-ascend 8f3fd59a2)复现,确认**非 w8a8 量化假象**,且新镜像 vllm/vllm-ascend 配对一致**无需任何源码 patch**。

### 启动脚本
```bash
export VLLM_VERSION=0.26.0          # 与镜像 vllm 版本一致
export MODEL_PATH=/home/admin/model-csi/models/modelhub_9000009_qwen3-235b-a22b-instruct-2507-103400091_20260116174303/model
vllm serve ${MODEL_PATH} \
  --host 0.0.0.0 --port 8000 --trust-remote-code --served-model-name auto \
  --distributed-executor-backend mp -tp=16 \
  --model-loader-extra-config '{"enable_multithread_load":true,"num_threads":8}' \
  --max-num-seqs 64 --max-num-batched-tokens 16384 --block-size 128 \
  --no-async-scheduling --gpu-memory-utilization 0.9 \
  --compilation-config '{"cudagraph_capture_sizes":[1,2,4,6,8,12,16,24,32,36,40,44,48,56,60,64,72,90,96,108,120,132,144],"cudagraph_mode":"FULL_DECODE_ONLY"}' \
  --structured-outputs-config '{"backend":"xgrammar","disable_any_whitespace":true}'
```
（BF16 235B=470G,TP=8 每卡 59G OOM,必须 TP=16;BF16 无 `--quantization ascend`）

### 复现 curl（宽松 schema,无 strict/无 maxItems）
```bash
curl -s -X POST http://<pod-ip>:8000/v1/chat/completions -H 'Content-Type: application/json' -d '{
  "model":"auto",
  "messages":[{"role":"user","content":"请生成一个包含多种类型字段的数据。"}],
  "response_format":{"type":"json_schema","json_schema":{"name":"multi_type_data","schema":{
    "type":"object","properties":{"name":{"type":"string"},"age":{"type":"integer"},
      "score":{"type":"number"},"is_active":{"type":"boolean"},
      "tags":{"type":"array","items":{"type":"string"}}},"required":["name","age"]}}},
  "stream":false,"max_tokens":4096}'
```

### 权重 / 镜像
- 模型: `qwen3-235b-a22b-instruct-2507`, **BF16**（无量化）
- modelhub: `modelhub_9000009_qwen3-235b-a22b-instruct-2507:103400091_20260116174303` (438G, 118 shards)
- 镜像: `hcr.meta-wulan01.hw-wulan.local/antsys/vllm:nightly-main-a3-openeuler-20260804110142_aarch64` (vllm 0.26.0 / vllm-ascend 8f3fd59a2, **无需 patch**)
- 硬件: 16x Ascend 910A3, TP=16

### 结果

**max_tokens=300, 宽松 schema, 5 发**: 4/5 = **80% 扩列**（1 发正常 20 item 停;4 发 finish=length,60-90 item）

| shot | finish | ctok | tag_items | expand |
|---|---|---|---|---|
| 1 | stop | 117 | 20 | ✗ |
| 2 | length | 300 | 60 | ✓ |
| 3 | length | 300 | 69 | ✓ |
| 4 | length | 300 | 84 | ✓ |
| 5 | length | 300 | 90 | ✓ |

**max_tokens=4096, 1 发**: `finish=length`, 4096 tokens, **999 tag items, 727 unique, 272 dup (27%)**。head/tail 全是不同标签（`"developer","python","data science",...` / `..."inconsistency","discrepancy","anomaly"...`），4096 token 仍在扩列阶段未到交替振荡。

**strict schema**(`strict:true`+全 required+`additionalProperties:false`, 3 发): 0/3 扩列（BF16+strict 不触发;对照 w8a8 strict 66%扩列）——**schema 宽松度是主导变量**,与 [[#44619]]/[[#17865]] 一致(maxItems/strict 才被 xgrammar 强制)。

### 结论
- array<string> 无 maxItems 扩列在 **BF16 无量化** + **最新 nightly-main 0.26.0** 上同样必现(宽松 schema 80%),**排除 w8a8 量化因素**,确认为 xgrammar grammar mask + 无界 array 的通用行为。
- `disable_any_whitespace=true` 对 array 无效(对 age integer 复读有效,5/5 正常)。
- 修复:schema 加 `maxItems`(已验证 0% 扩列)。

**Comment by QwertyJack** [comment]
We completed an end-to-end L0 reference with the same BF16 checkpoint using Transformers + torch CPU, with no `torch_npu`, vLLM, or vLLM-Ascend involved.

| Level | Path | Result at 300 tokens |
| --- | --- | --- |
| L0 | Transformers + torch CPU | `finish_reason=length`, 63 tags, 63 unique |
| L1 | Transformers + torch_npu EP16 | `finish_reason=length`, 69 tags, 69 unique |
| L2 | vLLM-Ascend TP16 | `finish_reason=length`, 92 tags, 17 unique |

The CPU run used the same prompt, tokenizer/chat template, schema, greedy decoding, and xgrammar 0.2.3 mask. At the frozen boundary, CPU and NPU have the same raw top-5 candidates and both prefer continuation over closure; vLLM-Ascend makes the same decision after masking.

The unbounded schema is therefore an enabling condition, not a sufficient cause: it permits continuation but does not force it. Other models or thinking variants close under the same schema, which further points to model-specific constrained-decoding behavior in `Qwen3-235B-A22B-Instruct-2507`. Since the same behavior reproduces end to end on CPU, the evidence does not support an Ascend hardware, `torch_npu`, or vLLM-Ascend inference-engine issue.

### Local environment

- Model: `Qwen3-235B-A22B-Instruct-2507`, BF16
- vLLM: `0.25.1`
- vLLM-Ascend: `0.19.1rc2.dev1265+g50e0ce608`
- Transformers: `5.14.1`
- torch / torch_npu: `2.10.0` / `2.10.0.post2`
- xgrammar: `0.2.3`
- CANN: `9.0.1` (`V100R001C10SPC002B220`)
- L0: aarch64 CPU, 320 OMP threads
- L1/L2 hardware: 16x Ascend 910; EP16 / TP16

**Comment by linnea-lin-00638949** [comment]
模型能力问题，客户gpu上也复现了，作为已知问题传递给业务方，客户在线系统有链路侧规避（截获结构化输出，在prompt内pad相关描述）
